test/npu-xrt/txn_basic/test.cpp

```diff
@@ -0,0 +1,254 @@
+                                 auto name = k.get_name();
+                                 std::cout << "Name: " << name << std::endl;
+                                 return name.rfind(Node, 0) == 0;
+                               });
+  auto kernelName = xkernel.get_name();
+
+  if (verbosity >= 1)
+    std::cout << "Registering xclbin: " << vm["xclbin"].as<std::string>()
+              << "\n";
+
+  device.register_xclbin(xclbin);
+
+  // get a hardware context
+  if (verbosity >= 1)
+    std::cout << "Getting hardware context." << std::endl;
+  xrt::hw_context context(device, xclbin.get_uuid());
+
+  // get a kernel handle
+  if (verbosity >= 1)
+    std::cout << "Getting handle to kernel:" << kernelName << std::endl;
+  auto kernel = xrt::kernel(context, kernelName);
+
+  auto bo_instr = xrt::bo(device, instr_v.size() * sizeof(int),
+                          XCL_BO_FLAGS_CACHEABLE, kernel.group_id(1));
+  auto bo_inA = xrt::bo(device, N * sizeof(int32_t), XRT_BO_FLAGS_HOST_ONLY,
+                        kernel.group_id(3));
+  auto bo_inB = xrt::bo(device, N * sizeof(int32_t), XRT_BO_FLAGS_HOST_ONLY,
+                        kernel.group_id(4));
+  auto bo_out = xrt::bo(device, N * sizeof(int32_t), XRT_BO_FLAGS_HOST_ONLY,
+                        kernel.group_id(5));
+
+  if (verbosity >= 1)
+    std::cout << "Writing data into buffer objects." << std::endl;
+
+  int32_t *bufInA = bo_inA.map<int32_t *>();
+  int32_t *bufInB = bo_inB.map<int32_t *>();
+  std::vector<uint32_t> srcVecA;
+  for (int i = 0; i < N; i++)
+    srcVecA.push_back(i + 1);
+
+  memcpy(bufInA, srcVecA.data(), (srcVecA.size() * sizeof(uint32_t)));
+  memcpy(bufInB, srcVecA.data(), (srcVecA.size() * sizeof(uint32_t)));
+
+  std::vector<uint32_t> outVec;
+  int32_t *bufOut = bo_out.map<int32_t *>();
+  for (int i = 0; i < N; i++)
+    outVec.push_back(i + 0x1011);
+  memcpy(bufOut, outVec.data(), (outVec.size() * sizeof(uint32_t)));
+
+  std::cout << "DDR AIE Address Offset: " << std::hex << DDR_AIE_ADDR_OFFSET
+            << std::endl;
+  std::cout << "bo_inA.address() + DDR_AIE_ADDR_OFFSET: " << std::hex
+            << bo_inA.address() + DDR_AIE_ADDR_OFFSET << std::endl;
+  std::cout << "bo_inB.address() + DDR_AIE_ADDR_OFFSET: " << std::hex
+            << bo_inB.address() + DDR_AIE_ADDR_OFFSET << std::endl;
+  std::cout << "bo_out.address() + DDR_AIE_ADDR_OFFSET: " << std::hex
+            << bo_out.address() + DDR_AIE_ADDR_OFFSET << std::endl;
+
+  std::vector<uint64_t> patches = {bo_out.address() + DDR_AIE_ADDR_OFFSET,
+                                   bo_inA.address() + DDR_AIE_ADDR_OFFSET};
+  auto p = patches.begin();
+  for (auto i = instr_v.begin(); i != instr_v.end(); i++) {
+    if (*i == 0xfeedbeef) {
+      uint64_t d = *p++;
+      *i = (d & 0xfffffffc);
+      *(i + 1) = (*(i + 1) & 0xFFFF0000) | ((d >> 32) & 0xffff);
+    }
+    std::cout << "Instruction: " << std::hex << std::setw(8)
+              << std::setfill('0') << *i << std::endl;
+  }
```

Comment:
Currently this is doing address patching manually, but there is also an instruction for this, which we haven't added to mlir-aie yet.

Reply:
> Currently this is doing address patching manually, but there is also an instruction for this, which we haven't added to mlir-aie yet.

This is now replaced by `aiex.npu.address_patch`